diagnostic_feedback/quiz.py

```diff
@@ -0,0 +1,436 @@
+    )
+
+    weight = Float(
+        display_name=_("Weight"),
+        help=_("Defines the maximum total grade of this question."),
+        default=1,
+        scope=Scope.content,
+        enforce_type=True
+    )
+
+    completed = Boolean(
+        default=False,
+        scope=Scope.user_state,
+        help=_("Has the student completed this quiz")
+    )
+
+    @property
+    def display_name_with_default(self):
+        return self.title
+
+    @property
+    def additional_publish_event_data(self):
+        return {
+            'user_id': self.scope_ids.user_id,
+            'block_id': self.get_block_id(),
+            'component_id': self.scope_ids.usage_id
+        }
+
+    has_score = True
+
+    def get_fragment(self, context, view='studio', json_args=None):
+        """
+        return fragment after loading css/js/html either for studio OR student view
+        :param context: context for templates
+        :param view: view_type i;e studio/student
+        :return: fragment after loading all assets
+        """
+        """
+            Return fragment after adding required css/js/html
+        """
+        fragment = Fragment()
+        self.add_templates(fragment, context, view)
+        self.add_css(fragment, view)
+        self.add_js(fragment, view)
+        self.initialize_js_classes(fragment, view, json_args)
+        return fragment
+
+    def append_choice(self, questions):
+        """
+        append student choice with each question if available
+        :param questions: list of questions
+        :return:
+        """
+        """
+
+        """
+        for question in questions:
+            if self.quiz_type == self.DIAGNOSTIC_QUIZ_VALUE:
+                question['student_choice'] = float(self.student_choices.get(question['id'])) if \
+                    self.student_choices.get(question['id']) else ''
+            else:
+                question['student_choice'] = self.student_choices.get(question['id'], '')
+
+    def get_block_id(self):
+        """
+        Return ID of `block`
+        """
+        usage_id = self.scope_ids.usage_id
+        # Try accessing block ID. If usage_id does not have it, return usage_id itself
+        return unicode(getattr(usage_id, 'block_id', usage_id))
+
+    def get_question(self, question_id):
+        """
+        Return Question object for given question id
+        """
+        question = {}
+        for question in self.questions:
+            if question['id'] == question_id:
+                question = question
+                break
+
+        return question
+
+    def get_buzzfeed_answer(self, choices, student_choice):
+        """
+        Return buzzfeed quiz answer label from question choices using student choice
+        """
+        choice_name = ''
+        for choice in choices:
+            if choice['category_id'] == student_choice:
+                choice_name = choice['name']
+                break
```

Comment:
Again, dict access will be faster.  This slows down as the number of choices increases.